test/test_smem_reuse.cpp

```diff
@@ -143,35 +144,247 @@ TEST_F(SmemReuseTest, NeedsReorderedPush) {
   auto fusion = std::make_unique<Fusion>();
   FusionGuard fg(fusion.get());
 
-  int64_t H_int = 5, W_int = 6;
-  auto H = IrBuilder::create<Val>(H_int);
-  auto W = IrBuilder::create<Val>(W_int);
+  int64_t H = 5;
 
-  auto tv0 = full({H}, fusion->oneVal(), DataType::Float);
-  auto tv1 = set(tv0); // pos = a. A = tv1
+  auto tv0 = full(
+      {IrBuilder::create<Val>(H)},
+      fusion->oneVal(),
+      DataType::Float); // pos = a. A = tv0
+  tv0->setMemoryType(MemoryType::Shared);
+
+  auto tv1 =
+      pad(tv0, {fusion->zeroVal(), fusion->oneVal()}); // pos = b. B = tv1
+  tv1->setMemoryType(MemoryType::Shared);
+
+  auto tv2 = mul(tv1, tv1); // pos = c
+
+  auto tv3 = sum(tv2, {0}); // gap between b and c. Can parallelize to sync
+
+  auto tv4 = broadcast(tv3, {true});
+  auto tv5 = mul(tv4, tv1); // pos = d. C = tv5
+  tv5->setMemoryType(MemoryType::Shared);
+
+  auto tv6 = add(tv1, tv1); // pos = e
+  fusion->addOutput(tv6);
+
+  auto tv7 = neg(tv5); // pos = f
+  fusion->addOutput(tv7);
+
+  { // This should not re-use memory
+    GpuLower gpulw(fusion.get());
+
+    ExpressionEvaluator ee;
+    std::unordered_set<int64_t> addresses;
+    int64_t smem_usage = 0;
+    for (auto alloc : gpulw.kernel()->summary().dynamic_smem_allocations) {
+      EXPECT_NE(alloc->address(), nullptr);
+      auto addr = ee.evaluate(alloc->address()).as<int64_t>();
+      TORCH_CHECK(
+          addresses.insert(addr).second,
+          "Smem addresses should not be re-used");
+      auto size = ee.evaluate(alloc->size()).as<int64_t>() *
+          dataTypeSize(alloc->buffer()->dtype());
+      smem_usage = std::max(smem_usage, addr + size);
+    }
+    EXPECT_EQ(
+        smem_usage, alignInt(alignInt((H + 1) * 4) + (H + 1) * 4) + H * 4);
+  }
+
+  { // Now introduce a block reduction and check that we re-use memory
+    tv3->axis(0)->parallelize(ParallelType::TIDx);
+
+    GpuLower gpulw(fusion.get());
+    ExpressionEvaluator ee;
+    int64_t smem_usage = 0;
+    for (auto alloc : gpulw.kernel()->summary().dynamic_smem_allocations) {
+      EXPECT_NE(alloc->address(), nullptr);
+      auto addr = ee.evaluate(alloc->address()).as<int64_t>();
+      auto size = ee.evaluate(alloc->size()).as<int64_t>() *
+          dataTypeSize(alloc->buffer()->dtype());
+      smem_usage = std::max(smem_usage, addr + size);
+    }
+    EXPECT_EQ(smem_usage, alignInt((H + 1) * 4) + (H + 1) * 4);
```

Comment:
Can we just directly validate the alias relationship of each allocation? I understand checking the total size is also fine, but asserting the alias relationships would make the intention of the test and expected behavior much more clear.

Reply:
Since they are not aliased (i.e. the `kir::Allocate`s have null `alias()`), we can't compare those directly. We can probably compare the addresses instead though. I will give it a shot.

Reply:
Oh, that's true. Then maybe not worth spending much time.